src/thumbnail.c

```diff
@@ -0,0 +1,242 @@
+{
+    const struct pixmap* full = &image->frames[0].pm;
+    const float scale_width = 1.0 / ((float)full->width / size);
+    const float scale_height = 1.0 / ((float)full->height / size);
+    const float scale =
+        fill ? max(scale_width, scale_height) : min(scale_width, scale_height);
+    size_t thumb_width = scale * full->width;
+    size_t thumb_height = scale * full->height;
+    ssize_t offset_x, offset_y;
+
+    if (fill) {
+        offset_x = size / 2 - thumb_width / 2;
+        offset_y = size / 2 - thumb_height / 2;
+        thumb_width = size;
+        thumb_height = size;
+    } else {
+        offset_x = 0;
+        offset_y = 0;
+    }
+
+    *params = (struct thumbnail_params) {
+        .thumb_width = thumb_width,
+        .thumb_height = thumb_height,
+        .offset_x = offset_x,
+        .offset_y = offset_y,
+        .fill = fill,
+        .antialias = antialias,
+        .scale = scale,
+    };
+}
+
+bool thumbnail_save(const struct pixmap* thumb, const char* source,
+                    const struct thumbnail_params* params)
+{
+    FILE* fp = NULL;
+    uint32_t i;
+    char* path = NULL;
+
+    if (!get_thumb_path(&path, source)) {
+        goto fail;
+    }
+
+    if (!make_directories(path)) {
+        goto fail;
+    }
+
+    if (!(fp = fopen(path, "wb"))) {
+        goto fail;
+    }
+
+    if (fprintf(fp, "P6\n%zu %zu\n255\n", thumb->width, thumb->height) < 0) {
+        goto fail;
+    }
+
+    /* comment to store params */
+    if (fputc('#', fp) == EOF ||
+        fwrite(params, sizeof(struct thumbnail_params), 1, fp) != 1 ||
+        fputc('\n', fp) == EOF) {
+        goto fail;
+    }
+
+    // TODO: add alpha channel
+    for (i = 0; i < thumb->width * thumb->height; ++i) {
+        uint8_t color[] = { (((thumb->data[i] >> (8 * 2)) & 0xff)),
+                            (((thumb->data[i] >> (8 * 1)) & 0xff)),
+                            (((thumb->data[i] >> (8 * 0)) & 0xff)) };
+        fwrite(color, 3, 1, fp);
+    }
+
+    free(path);
+    fclose(fp);
+    return true;
+
+fail:
+    free(path);
+    if (fp) {
+        fclose(fp);
+    }
+    return false;
+}
+
+/**
+ * Allocate/reallocate pixel map.
+ * @param thumb pixmap context
+ * @param path path to load from
+ * @return true pixmap was loaded
+ */
+bool thumbnail_load(struct pixmap* thumb, const char* source,
+                    const struct thumbnail_params* params)
+{
+    FILE* fp = NULL;
+    char* path = NULL;
+    uint32_t i;
+    struct stat attr_img, attr_thumb;
+    struct thumbnail_params saved_params;
+
+    if (!get_thumb_path(&path, source) || stat(source, &attr_img) ||
+        stat(path, &attr_thumb) ||
+        difftime(attr_img.st_ctime, attr_thumb.st_ctime) > 0) {
+        goto fail;
+    }
+
+    if (!(fp = fopen(path, "rb"))) {
+        goto fail;
+    }
+
+    char header[3];
+    if (fscanf(fp, "%2s\n%zu %zu\n255\n", header, &thumb->width,
+               &thumb->height) != 3) {
+        goto fail;
```

Comment:
You love `goto` so much =)
Maybe there is another way to handle errors?

Reply:
This is the same way of handling errors as in `formats/avif.c` and many other files in the project. I could maybe make a macro which cleans stuff up and then returns false, but I believe this is cleaner.